### ISSUE TITLE
Migration helpers wrapped in blocks always run in 'up' direction

### ISSUE THREAD

**Issue by Sinjo** [rails/rails#21484]
When using a helper which is designed to wrap other migration steps in a block, those steps always run in the `up` direction. This is reproducible with a very simple migration (which I've included in a [sample Rails project](https://github.com/Sinjo/activerecord_migration_block_repro)):

``` ruby
class TransactionBlockWontReverse < ActiveRecord::Migration
  def change
    transaction do
      create_table :users do |t|
        t.column :name, :string
      end
    end
  end
end
```

`rake db:migrate` works fine for this migration, but `rake db:rollback` exits with an exception:

```
$ rake db:rollback
== 20150902142918 TransactionBlockWontReverse: reverting ======================
-- transaction()
-- create_table(:users)
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

PG::DuplicateTable: ERROR:  relation "users" already exists
: CREATE TABLE "users" ("id" serial primary key, "name" character varying) /Users/sinjo/projects/rails/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:155:in `async_exec'
```

I first found this on a project using Rails 4.2, but it happens on master too. I'm happy to work on a fix, but I thought I'd raise it here to see if any core devs have opinions on how this should work.

**Comment by al2o3cr** [comment]
Looks like reversing `transaction` just plays back whatever's in the block:

https://github.com/rails/rails/blob/314eb6596a36732d907c5cd115f5262904d13dad/activerecord/lib/active_record/migration/command_recorder.rb#L110

My guess is that it should actually be wrapping the passed block in `revert` so that it also runs backwards.

**Comment by Sinjo** [comment]
As far as I can tell, `revert` returns an array of the inverse commands for everything run in the block passed to it. I can't see how you could use that in the return value of `invert_transaction`, but I'm probably missing something. Could you clarify how you'd use it?

**Comment by senny** [comment]
> When using a helper which is designed to wrap other migration steps in a block

You're example uses `transaction` but what kind of helper do you use in your production code? Or asked differently, does the patch from @dv, which is specific to transactions solve your issue?

**Comment by senny** [comment]
@Sinjo would you mind having a look at #21484 and let me know wether that is sufficient to solve your issue?

**Comment by Sinjo** [comment]
I was working on a new helper and made the (incorrect) assumption that blocks would automatically be reversed by `CommandRecorder`. I'd implemented it in a similar style to `transaction` and jumped to conclusions about how blocks should work.

Sorry for the confusion! Awesome to see a fix in for `transaction` though. :)

**Comment by y-yagi** [comment]
Fixed by #31604.
